c10/util/order_preserving_flat_hash_map.h

```diff
@@ -678,48 +677,71 @@ class sherwood_v3_table : private EntryAlloc, private Hasher, private Equal
             rehash(required_buckets);
     }
 
+    void replace_linked_list_position(EntryPointer to_be_replaced, EntryPointer new_node) {
+      remove_from_list(new_node);
+      insert_after(new_node, to_be_replaced->prev);
+      remove_from_list(to_be_replaced);
+    }
+
     // the return value is a type that can be converted to an iterator
     // the reason for doing this is that it's not free to find the
     // iterator pointing at the next element. if you care about the
     // next iterator, turn the return value into an iterator
     convertible_to_iterator erase(const_iterator to_erase)
     {
         EntryPointer current = to_erase.current;
+        remove_from_list(current);
         current->destroy_value();
         --num_elements;
+
         for (EntryPointer next = current + ptrdiff_t(1); !next->is_at_desired_position(); ++current, ++next)
         {
+            // if an entry is being removed, and there are other entries with the
+            // same hash, the other entries get moved to their desired position by
+            // reinserting.
             current->emplace(next->distance_from_desired - 1, std::move(next->value));
+            replace_linked_list_position(next, current);
             next->destroy_value();
         }
         return { to_erase.current };
     }
 
     iterator erase(const_iterator begin_it, const_iterator end_it)
     {
-        if (begin_it == end_it)
-            return { begin_it.current };
-        for (EntryPointer it = begin_it.current, end = end_it.current; it != end; ++it)
-        {
-            if (it->has_value())
+        // whenever an entry is removed and there are other entries with the same
+        // hash, the other entries must get moved to their desired position.
+        // any reference to a moved entry is invalidated.
+        // here, we iterate through the range, and make sure that we update
+        // the pointer to our next entry in the list or the end of the iterator
```

Comment:
Why does this `erase` have so many re-writes from the original? Can't it do the same thing as the other `erase` and restore the positions in the list after the `emplace`?

Reply:
In the original implementation, the erase seems to exploit the fact that iterating entries in order was efficient and led to efficient moves. When no longer doing this, it seems like we can just use the single element  `erase` in a loop:
```
while(it != end) {
  auto cur = it++;
  erase(cur);
}
```

Reply:
`end` might get invalidated in the single element erase